File: delta/src/drift.js
// Детектор соціального дрейфу — приватне дзеркало (M5, docs/specs/
// 260809-delta-app.md, «Обсяг M5», п.4; mt: mandates.md, «Детектор
// соціального дрейфу»: «класи рішень, де людські вибори систематично
// відхиляються від мети задачі в бік соціального комфорту... показуються
// власнику мандата як дельта «мета vs комфорт»... адресовано самому
// власнику, не нагляд згори»; конституція п.6: «приходить лише самому
// власнику, з кнопкою «делегувати цю дію ШІ»»).
//
// **Headless-actor патерн** — той самий, що `watcher.js`: `driftScan` —
// pure-функція над скановими `decisionsDirs` (не читає диск сама),
// `runDriftScan` — оркеструє скан + персист, спільна точка входу для CLI/GUI
// tool `drift_scan` і майбутнього headless-крону (dogfooding з watcher-ом).
//
// **Сигнали систематичного відкладання** (клас = `decision_type`,
// одноосібні рішення — `requiresQuorum` кворумні виключено: кворум має
// власну механіку прогресу, не «моє особисте відкладання»):
//  (а) **застарілі** — відкриті (немає `NNNN-approval.json`) старші за
//      `staleDays` (дефолт 7 — власний поріг M5, задокументоване рішення:
//      м'якший за watcher-івський SLA+grace 24+24h, дрейф — довший
//      горизонт «я системно уникаю цього класу», не «це щойно висить»);
//  (б) **повторні ітерації без підпису** — квіз-файл (`NNNN-quiz.md`)
//      існує з `iterations >= iterationsThreshold` (дефолт 3), а рішення
//      й досі відкрите — «я почав, провалив кілька спроб і покинув», не
//      «я ще не почав».
// Обидва сигнали ОКРЕМІ (задокументоване спрощення M5): «reject-и з
// подальшим поверненням того самого класу» (третій сигнал з задачі)
// потребує графової прив'язки причин-наслідків між закритими й новими
// рішеннями, якої файловий мок не матеріалізує (той самий чесно названий
// ліміт, що `track-record.js: override` — «не обов'язково та сама
// розвилка»); кандидат наступного мілстоуна, коли precedent-engine
// (mandates.md, «Прецедентний рушій») дасть графову прив'язку.
//
// **Приватність — ЛОКАЛЬНО поза git, НЕ в `.mt/notifications`** (буквально
// з задачі): картки пишуться біля `knowledge.json`/`device_key.json`
// (`drift.json`, той самий рівень приватності — файл-сусід `config.json`
// цього пристрою), НЕ у спільний `.mt/notifications/{handle}.jsonl`
// watcher-а (той лог читають інші акти системи для ескалації — дрейф
// нікуди не ескалює, це дзеркало самому собі).

import { parseDecisionRequest } from './decisions.js'

const DECISION_REQUEST_RE = /^(\d{4})-decision-request\.md$/
const DAY_MS = 24 * 60 * 60 * 1000
const DEFAULT_STALE_DAYS = 7
const DEFAULT_ITERATIONS_THRESHOLD = 3

/**
 * @returns {{staleDays: number, iterationsThreshold: number}} дефолтний конфіг детектора
 */
export function defaultDriftConfig() {
  return { staleDays: DEFAULT_STALE_DAYS, iterationsThreshold: DEFAULT_ITERATIONS_THRESHOLD }
}

/**
 * @param {string} openedAtIso ISO-час відкриття decision-request
 * @param {Date} now поточний час
 * @returns {number|null} вік у днях, або `null` — `openedAtIso` невалідний/відсутній
 */
function ageDaysOf(openedAtIso, now) {
  if (!openedAtIso) return null
  const openedMs = Date.parse(openedAtIso)
  if (Number.isNaN(openedMs)) return null
  return (now.getTime() - openedMs) / DAY_MS
}

const ITERATIONS_FIELD_RE = /^iterations: (\d+)$/m

/**
 * `iterations` квіз-чернетки цього `nnnn`, якщо файл є — читає СИРИЙ
 * `iterations:` рядок фронтматера напряму (без повного `quiz.js: parseQuizFile`
 * — drift.js навмисно не залежить від формату Q&A/teach-back квіз-файлів,
 * лише від спільного для ОБОХ поля `iterations`, той самий інваріант, що
 * обидва формати документують у заголовку: «iterations — сумарна кількість
 * спроб»).
 * @param {string|undefined} quizText сирий вміст `NNNN-quiz.md`, якщо є
 * @returns {number} `iterations`, або `0` — квізу немає/поле відсутнє
 */
function quizIterations(quizText) {
  if (!quizText) return 0
  const match = ITERATIONS_FIELD_RE.exec(quizText)
  return match ? Number(match[1]) : 0
}

/**
 * Деривує дрейф-картки одного власника з уже сканованих `decisions/`-
 * директорій — pure-функція (той самий підхід, що `watcher.js:
 * scanForNotifications`): не читає диск сама, `now` завжди явно
 * ін'єктований.
 * @param {object} params вхідні параметри
 * @param {{dir: string, files: {name: string, content: string}[]}[]} params.decisionsDirs скановані `decisions/`-директорії
 * @param {string|null|undefined} params.handle власник, чий дрейф деривувати
 * @param {Date} params.now поточний час (ін'єкція годинника)
 * @param {number} [params.staleDays] поріг «застаріле» в днях
 * @param {number} [params.iterationsThreshold] поріг «повторні ітерації без підпису»
 * @returns {{decisionType: string, count: number, oldestAgeDays: number, deadlineCostSample: string|null,
 *   items: {nnnn: string, runId: string|null, ageDays: number|null, iterations: number, deadlineCost: string|null,
 *   signal: 'stale'|'repeated-iterations'|'both'}[], generatedAt: string}[]} дрейф-картки, найстаріший клас перший
 */
export function detectDrift({ decisionsDirs, handle, now, staleDays = DEFAULT_STALE_DAYS, iterationsThreshold = DEFAULT_ITERATIONS_THRESHOLD }) {
  if (!handle) return []
  const items = collectDriftItems({ decisionsDirs, handle, now, staleDays, iterationsThreshold })

  const byType = new Map()
  for (const item of items) {
    if (!byType.has(item.decisionType)) byType.set(item.decisionType, [])
    byType.get(item.decisionType).push(item)
  }

  return Array.from(byType, ([decisionType, typeItems]) => {
    const sorted = typeItems.toSorted((a, b) => (b.ageDays ?? 0) - (a.ageDays ?? 0))
    return {
      decisionType,
      count: sorted.length,
      oldestAgeDays: Math.round(Math.max(...sorted.map(i => i.ageDays ?? 0))),
      deadlineCostSample: sorted.find(i => i.deadlineCost)?.deadlineCost ?? null,
      items: sorted.map(({ decisionType: _dt, ...rest }) => rest),
      generatedAt: now.toISOString()
    }
  }).toSorted((a, b) => b.oldestAgeDays - a.oldestAgeDays)
}

/**
 * @param {object} params вхідні параметри
 * @param {{dir: string, files: {name: string, content: string}[]}[]} params.decisionsDirs скановані `decisions/`-директорії
 * @param {string|null|undefined} params.handle власник, чий дрейф деривувати
 * @param {Date} params.now поточний час
 * @param {number} params.staleDays поріг «застаріле» в днях
 * @param {number} params.iterationsThreshold поріг «повторні ітерації без підпису»
 * @returns {object[]} плаский список сигнальних item-ів (з `decisionType` для групування)
 */
function collectDriftItems({ decisionsDirs, handle, now, staleDays, iterationsThreshold }) {
  const items = []
  for (const { dir, files } of decisionsDirs ?? []) {
    const filesByName = new Map(files.map(f => [f.name, f.content]))
    for (const file of files) {
      const match = DECISION_REQUEST_RE.exec(file.name)
      if (!match) continue
      const nnnn = match[1]
      const item = driftItemOrNull({ file, filesByName, nnnn, dir, handle, now, staleDays, iterationsThreshold })
      if (item) items.push(item)
    }
  }
  return items
}

/**
 * @param {object} params вхідні параметри одного decision-request-кандидата
 * @param {{name: string, content: string}} params.file файл decision-request-а
 * @param {Map<string, string>} params.filesByName карта імʼя файлу → вміст у директорії decisions
 * @param {string} params.nnnn чотиризначний номер
 * @param {string} params.dir абсолютний шлях до decisions-директорії
 * @param {string} params.handle власник, чий дрейф деривувати
 * @param {Date} params.now поточний час
 * @param {number} params.staleDays поріг «застаріле» в днях
 * @param {number} params.iterationsThreshold поріг «повторні ітерації без підпису»
 * @returns {object|null} сигнальний item, або `null` — не мій/кворумний/закритий/без сигналу
 */
function driftItemOrNull({ file, filesByName, nnnn, dir, handle, now, staleDays, iterationsThreshold }) {
  const parsed = parseDecisionRequest(file.content, { path: `${dir}/${file.name}`, nnnn })
  if (parsed.computedOwner !== handle) return null
  if (parsed.leverageFacets.irreversible) return null // кворум — власна механіка прогресу, не «моє особисте відкладання»
  if (filesByName.has(`${nnnn}-approval.json`)) return null // закрите — не відкладається

  const age = ageDaysOf(parsed.openedAt, now)
  const iterations = quizIterations(filesByName.get(`${nnnn}-quiz.md`))
  const stale = age !== null && age >= staleDays
  const repeated = iterations >= iterationsThreshold
  if (!stale && !repeated) return null

  const runId = dir.split('/').filter(Boolean).at(-2) ?? null
  return {
    decisionType: parsed.decisionType ?? 'general',
    nnnn,
    runId,
    ageDays: age === null ? null : Math.round(age),
    iterations,
    deadlineCost: parsed.deadlineCost,
    signal: driftSignal(stale, repeated)
  }
}

/**
 * @param {boolean} stale чи спрацював сигнал «застаріле»
 * @param {boolean} repeated чи спрацював сигнал «повторні ітерації без підпису»
 * @returns {'stale'|'repeated-iterations'|'both'} комбінований сигнал
 */
function driftSignal(stale, repeated) {
  if (stale && repeated) return 'both'
  return stale ? 'stale' : 'repeated-iterations'
}

/**
 * Розбирає локальний файл дрейф-карток (`drift.json`, поза git) — той самий
 * fail-safe парсинг, що `knowledge.js: parseKnowledgeFile`.
 * @param {string|null|undefined} text сирий вміст `drift.json`
 * @returns {object[]} дрейф-картки
 */
export function parseDriftFile(text) {
  if (!text) return []
  try {
    const parsed = JSON.parse(text)
    return Array.isArray(parsed) ? parsed : []
  } catch {
    return []
  }
}

/**
 * @param {object[]} cards дрейф-картки
 * @returns {string} pretty-print JSON текст файлу
 */
export function formatDriftFile(cards) {
  return `${JSON.stringify(cards, null, 2)}\n`
}

/**
 * Читає ЛОКАЛЬНІ дрейф-картки (той самий `{read, write}`-io-контракт, що
 * `knowledge.js: loadKnowledgeEntries`).
 * @param {{read: () => Promise<string|null>}|null|undefined} driftIo io дрейф-файлу
 * @returns {Promise<object[]>} дрейф-картки
 */
export async function loadDriftCards(driftIo) {
  if (!driftIo) return []
  return parseDriftFile(await driftIo.read())
}

/**
 * Перезаписує ЛОКАЛЬНІ дрейф-картки СВІЖИМ результатом скану (кожен
 * `drift_scan` — повний перерахунок, не інкрементальний append — той самий
 * підхід, що `watcher.js` НЕ використовує тут: watcher дописує, дрейф
 * перераховує, бо картка «клас Х відкладається» застаріває сама собою,
 * коли клас закрито — append накопичував би мертві картки назавжди).
 * @param {{write: (content: string) => Promise<void>}|null|undefined} driftIo io дрейф-файлу
 * @param {object[]} cards свіжі дрейф-картки
 * @returns {Promise<void>}
 */
export async function saveDriftCards(driftIo, cards) {
  if (!driftIo) return
  await driftIo.write(formatDriftFile(cards))
}

/**
 * Повний потік `drift_scan`: деривує картки ({@link detectDrift}) і
 * перезаписує локальний файл ({@link saveDriftCards}) — спільна точка
 * входу для CLI/GUI tool і headless-крону.
 * @param {object} params вхідні параметри
 * @param {{dir: string, files: {name: string, content: string}[]}[]} params.decisionsDirs скановані `decisions/`-директорії
 * @param {string|null|undefined} params.handle власник
 * @param {{read: () => Promise<string|null>, write: (content: string) => Promise<void>}} [params.driftIo] io дрейф-файлу
 * @param {{staleDays: number, iterationsThreshold: number}} [params.config] конфіг детектора
 * @param {() => Date} [params.now] ін'єкція годинника (тести)
 * @returns {Promise<object[]>} свіжі дрейф-картки
 */
export async function runDriftScan({ decisionsDirs, handle, driftIo, config, now }) {
  const resolvedConfig = { ...defaultDriftConfig(), ...config }
  const nowDate = now ? now() : new Date()
  const cards = detectDrift({ decisionsDirs, handle, now: nowDate, ...resolvedConfig })
  await saveDriftCards(driftIo, cards)
  return cards
}

File: delta/crates/delta-core/src/drift.rs
//! Детектор соціального дрейфу — приватне дзеркало, порт `delta/src/drift.js`
//! (M5). Headless-actor патерн, той самий, що `watcher`: `detect_drift` —
//! pure-функція над сканованими `decisionsDirs`, `run_drift_scan` —
//! оркеструє скан + персист. Картки — ЛОКАЛЬНО поза git (`drift.json`),
//! НЕ в спільний `.mt/notifications`.

use regex::Regex;
use serde::Serialize;

use crate::decisions::{parse_decision_request, DecisionsDir};
use crate::io::KnowledgeIo;

const DEFAULT_STALE_DAYS: f64 = 7.0;
const DEFAULT_ITERATIONS_THRESHOLD: u64 = 3;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct DriftConfig {
    pub stale_days: f64,
    pub iterations_threshold: u64,
}

pub fn default_drift_config() -> DriftConfig {
    DriftConfig {
        stale_days: DEFAULT_STALE_DAYS,
        iterations_threshold: DEFAULT_ITERATIONS_THRESHOLD,
    }
}

fn age_days_of(opened_at_iso: Option<&str>, now: chrono::DateTime<chrono::Utc>) -> Option<f64> {
    let opened = chrono::DateTime::parse_from_rfc3339(opened_at_iso?).ok()?;
    Some(
        (now.timestamp_millis() - opened.timestamp_millis()) as f64 / (24.0 * 60.0 * 60.0 * 1000.0),
    )
}

/// `iterations` квіз-чернетки, якщо файл є — читає сирий `iterations:`
/// рядок фронтматера напряму (`drift.js: quizIterations`).
fn quiz_iterations(quiz_text: Option<&str>) -> u64 {
    let Some(text) = quiz_text else { return 0 };
    let re = Regex::new(r"(?m)^iterations: (\d+)$").unwrap();
    re.captures(text)
        .and_then(|c| c[1].parse().ok())
        .unwrap_or(0)
}

#[derive(Debug, Clone, PartialEq, Serialize, serde::Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DriftSignal {
    Stale,
    RepeatedIterations,
    Both,
}

#[derive(Debug, Clone, PartialEq, Serialize, serde::Deserialize)]
pub struct DriftItem {
    pub nnnn: String,
    #[serde(rename = "runId")]
    pub run_id: Option<String>,
    #[serde(rename = "ageDays")]
    pub age_days: Option<i64>,
    pub iterations: u64,
    #[serde(rename = "deadlineCost")]
    pub deadline_cost: Option<String>,
    pub signal: DriftSignal,
}

#[derive(Debug, Clone, PartialEq, Serialize, serde::Deserialize)]
pub struct DriftCard {
    #[serde(rename = "decisionType")]
    pub decision_type: String,
    pub count: usize,
    #[serde(rename = "oldestAgeDays")]
    pub oldest_age_days: i64,
    #[serde(rename = "deadlineCostSample")]
    pub deadline_cost_sample: Option<String>,
    pub items: Vec<DriftItem>,
    #[serde(rename = "generatedAt")]
    pub generated_at: String,
}

struct RawItem {
    decision_type: String,
    item: DriftItem,
}

fn drift_signal(stale: bool, repeated: bool) -> DriftSignal {
    if stale && repeated {
        DriftSignal::Both
    } else if stale {
        DriftSignal::Stale
    } else {
        DriftSignal::RepeatedIterations
    }
}

#[allow(clippy::too_many_arguments)]
fn drift_item_or_none(
    name: &str,
    content: &str,
    files_by_name: &std::collections::HashMap<String, String>,
    nnnn: &str,
    dir: &str,
    handle: &str,
    now: chrono::DateTime<chrono::Utc>,
    config: &DriftConfig,
) -> Option<RawItem> {
    let parsed = parse_decision_request(
        content,
        crate::decisions::DecisionRequestMeta {
            path: Some(format!("{dir}/{name}")),
            nnnn: Some(nnnn.to_string()),
            ..Default::default()
        },
    )
    .ok()?;
    if parsed.computed_owner.as_deref() != Some(handle) {
        return None;
    }
    if parsed.leverage_facets.irreversible {
        return None;
    }
    if files_by_name.contains_key(&format!("{nnnn}-approval.json")) {
        return None;
    }

    let age = age_days_of(parsed.opened_at.as_deref(), now);
    let iterations = quiz_iterations(
        files_by_name
            .get(&format!("{nnnn}-quiz.md"))
            .map(|s| s.as_str()),
    );
    let stale = age.is_some_and(|a| a >= config.stale_days);
    let repeated = iterations >= config.iterations_threshold;
    if !stale && !repeated {
        return None;
    }

    let run_id = dir
        .split('/')
        .filter(|s| !s.is_empty())
        .rev()
        .nth(1)
        .map(str::to_string);
    Some(RawItem {
        decision_type: parsed
            .decision_type
            .unwrap_or_else(|| "general".to_string()),
        item: DriftItem {
            nnnn: nnnn.to_string(),
            run_id,
            age_days: age.map(|a| a.round() as i64),
            iterations,
            deadline_cost: parsed.deadline_cost,
            signal: drift_signal(stale, repeated),
        },
    })
}

/// Деривує дрейф-картки одного власника — pure-функція (`drift.js:
/// detectDrift`).
pub fn detect_drift(
    decisions_dirs: &[DecisionsDir],
    handle: Option<&str>,
    now: chrono::DateTime<chrono::Utc>,
    config: Option<DriftConfig>,
) -> Vec<DriftCard> {
    let Some(handle) = handle else {
        return Vec::new();
    };
    let config = config.unwrap_or_else(default_drift_config);
    let re = Regex::new(r"^(\d{4})-decision-request\.md$").unwrap();

    let mut items: Vec<RawItem> = Vec::new();
    for d in decisions_dirs {
        let files_by_name: std::collections::HashMap<String, String> =
            d.files.iter().cloned().collect();
        for (name, content) in &d.files {
            let Some(caps) = re.captures(name) else {
                continue;
            };
            let nnnn = &caps[1];
            if let Some(item) = drift_item_or_none(
                name,
                content,
                &files_by_name,
                nnnn,
                &d.dir,
                handle,
                now,
                &config,
            ) {
                items.push(item);
            }
        }
    }

    let mut by_type: std::collections::HashMap<String, Vec<DriftItem>> =
        std::collections::HashMap::new();
    for raw in items {
        by_type.entry(raw.decision_type).or_default().push(raw.item);
    }

    let mut cards: Vec<DriftCard> = by_type
        .into_iter()
        .map(|(decision_type, mut typed_items)| {
            typed_items.sort_by_key(|b| std::cmp::Reverse(b.age_days.unwrap_or(0)));
            let oldest = typed_items
                .iter()
                .map(|i| i.age_days.unwrap_or(0))
                .max()
                .unwrap_or(0);
            let sample = typed_items.iter().find_map(|i| i.deadline_cost.clone());
            DriftCard {
                decision_type,
                count: typed_items.len(),
                oldest_age_days: oldest,
                deadline_cost_sample: sample,
                items: typed_items,
                generated_at: now.to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
            }
        })
        .collect();
    cards.sort_by_key(|b| std::cmp::Reverse(b.oldest_age_days));
    cards
}

/// Розбирає локальний файл дрейф-карток — fail-safe (`drift.js:
/// parseDriftFile`).
pub fn parse_drift_file(text: Option<&str>) -> Vec<DriftCard> {
    let Some(text) = text else { return Vec::new() };
    serde_json::from_str(text).unwrap_or_default()
}

pub fn format_drift_file(cards: &[DriftCard]) -> String {
    format!(
        "{}\n",
        serde_json::to_string_pretty(cards).expect("Vec<DriftCard> серіалізується без помилок")
    )
}

pub async fn load_drift_cards(drift_io: Option<&dyn KnowledgeIo>) -> Vec<DriftCard> {
    let Some(io) = drift_io else {
        return Vec::new();
    };
    parse_drift_file(io.read().await.as_deref())
}

pub async fn save_drift_cards(drift_io: Option<&dyn KnowledgeIo>, cards: &[DriftCard]) {
    let Some(io) = drift_io else { return };
    io.write(&format_drift_file(cards)).await;
}

/// Повний потік `drift_scan`: деривує картки і ПЕРЕЗАПИСУЄ локальний файл
/// (не append — застаріла картка зникає сама, коли клас закрито) —
/// `drift.js: runDriftScan`.
pub async fn run_drift_scan(
    decisions_dirs: &[DecisionsDir],
    handle: Option<&str>,
    drift_io: Option<&dyn KnowledgeIo>,
    config: Option<DriftConfig>,
    now: chrono::DateTime<chrono::Utc>,
) -> Vec<DriftCard> {
    let cards = detect_drift(decisions_dirs, handle, now, config);
    save_drift_cards(drift_io, &cards).await;
    cards
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::io::MemoryKnowledgeIo;

    const DR_0004: &str =
        include_str!("../tests/fixtures/runs/demo-1/decisions/0004-decision-request.md");
    const DR_0001: &str =
        include_str!("../tests/fixtures/runs/demo-1/decisions/0001-decision-request.md");
    const DR_IRREVERSIBLE: &str =
        include_str!("../tests/fixtures/runs/demo-5/decisions/0001-decision-request.md");

    fn now() -> chrono::DateTime<chrono::Utc> {
        chrono::DateTime::parse_from_rfc3339("2026-08-09T10:00:00.000Z")
            .unwrap()
            .with_timezone(&chrono::Utc)
    }

    #[test]
    fn default_config_matches_spec() {
        let c = default_drift_config();
        assert_eq!(c.stale_days, 7.0);
        assert_eq!(c.iterations_threshold, 3);
    }

    #[test]
    fn no_handle_is_empty() {
        assert!(detect_drift(&[], None, now(), None).is_empty());
    }

    #[test]
    fn stale_ops_decision_is_one_card() {
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![("0004-decision-request.md".into(), DR_0004.into())],
        }];
        let cards = detect_drift(&dirs, Some("olena"), now(), None);
        assert_eq!(cards.len(), 1);
        assert_eq!(cards[0].decision_type, "ops");
        assert_eq!(cards[0].count, 1);
        assert_eq!(cards[0].items[0].signal, DriftSignal::Stale);
        assert!(cards[0].items[0].age_days.unwrap() >= 39);
        assert_eq!(
            cards[0].deadline_cost_sample.as_deref(),
            Some("немає залежних задач")
        );
    }

    #[test]
    fn fresh_open_decision_not_drift() {
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![("0001-decision-request.md".into(), DR_0001.into())],
        }];
        assert!(detect_drift(&dirs, Some("olena"), now(), None).is_empty());
    }

    #[test]
    fn closed_decision_not_drift_even_if_stale() {
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![
                ("0004-decision-request.md".into(), DR_0004.into()),
                ("0004-approval.json".into(), r#"{"approved":true}"#.into()),
            ],
        }];
        assert!(detect_drift(&dirs, Some("olena"), now(), None).is_empty());
    }

    #[test]
    fn other_owner_not_drift() {
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![("0004-decision-request.md".into(), DR_0004.into())],
        }];
        assert!(detect_drift(&dirs, Some("vitalii"), now(), None).is_empty());
    }

    #[test]
    fn irreversible_not_drift() {
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-5/decisions".into(),
            files: vec![("0001-decision-request.md".into(), DR_IRREVERSIBLE.into())],
        }];
        assert!(detect_drift(&dirs, Some("vitalii"), now(), None).is_empty());
    }

    #[test]
    fn repeated_iterations_signal_even_if_fresh() {
        let quiz_text = "---\nschema_version: 1\ndepth: one-tap\niterations: 5\n---\n";
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![
                ("0001-decision-request.md".into(), DR_0001.into()),
                ("0001-quiz.md".into(), quiz_text.into()),
            ],
        }];
        let cards = detect_drift(&dirs, Some("olena"), now(), None);
        assert_eq!(cards.len(), 1);
        assert_eq!(cards[0].items[0].signal, DriftSignal::RepeatedIterations);
        assert_eq!(cards[0].items[0].iterations, 5);
    }

    #[test]
    fn custom_thresholds_apply() {
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![("0004-decision-request.md".into(), DR_0004.into())],
        }];
        let config = DriftConfig {
            stale_days: 1000.0,
            iterations_threshold: 3,
        };
        assert!(detect_drift(&dirs, Some("olena"), now(), Some(config)).is_empty());
    }

    #[test]
    fn format_parse_round_trip() {
        let cards = vec![DriftCard {
            decision_type: "ops".into(),
            count: 1,
            oldest_age_days: 39,
            deadline_cost_sample: None,
            items: vec![],
            generated_at: now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        }];
        assert_eq!(parse_drift_file(Some(&format_drift_file(&cards))), cards);
    }

    #[test]
    fn parse_missing_or_corrupt_is_empty() {
        assert!(parse_drift_file(None).is_empty());
        assert!(parse_drift_file(Some("not json")).is_empty());
    }

    #[tokio::test]
    async fn load_save_without_io_are_noop() {
        assert!(load_drift_cards(None).await.is_empty());
        save_drift_cards(None, &[]).await;
    }

    #[tokio::test]
    async fn run_drift_scan_overwrites_not_appends() {
        let stale_card = DriftCard {
            decision_type: "stale-class".into(),
            count: 99,
            oldest_age_days: 999,
            deadline_cost_sample: None,
            items: vec![],
            generated_at: "2020-01-01T00:00:00.000Z".into(),
        };
        let io = MemoryKnowledgeIo::new(Some(format_drift_file(&[stale_card])));
        let dirs = vec![DecisionsDir {
            dir: "/root/runs/demo-1/decisions".into(),
            files: vec![("0004-decision-request.md".into(), DR_0004.into())],
        }];
        let cards = run_drift_scan(&dirs, Some("olena"), Some(&io), None, now()).await;
        assert_eq!(cards.len(), 1);
        assert_eq!(cards[0].decision_type, "ops");
        let persisted = load_drift_cards(Some(&io)).await;
        assert_eq!(persisted, cards);
        assert!(!persisted.iter().any(|c| c.decision_type == "stale-class"));
    }
}
